File: Trees/BinaryTrees/ViewsOfTree.java
package Trees.BinaryTrees;

import java.util.*;

public class ViewsOfTree {
    private static void findRightView(TreeNode root) {
        Map<Integer, ArrayList<Integer>> lookup = new HashMap();
        Queue<Pair> queue = new LinkedList<>();
        queue.add(new Pair(root, 0));
        while (!queue.isEmpty()) {
            Pair currentPair = queue.poll();
            TreeNode currentNode = currentPair.root;
            int currentNodeLevel = currentPair.stateCount;

            if (lookup.containsKey(currentNodeLevel)) {
                ArrayList<Integer> arr = lookup.get(currentNodeLevel);
                arr.add(currentNode.data);
                lookup.put(currentNodeLevel, arr);
            } else {
                ArrayList<Integer> arr = new ArrayList<>();
                arr.add(currentNode.data);
                lookup.put(currentNodeLevel, arr);
            }

            if (currentNode.left != null) queue.add(new Pair(currentNode.left, currentNodeLevel + 1));
            if (currentNode.right != null) queue.add(new Pair(currentNode.right, currentNodeLevel + 1));
        }

        for (Map.Entry<Integer, ArrayList<Integer>> map : lookup.entrySet()) {
            ArrayList<Integer> arrayList = map.getValue();
            int lastItem = arrayList.get(arrayList.size() - 1);
            System.out.print(lastItem + " ");
        }

    }

    private static void findRightViewOptimized(TreeNode root) {
        Map<Integer, Integer> lookups = new HashMap<>();
        Queue<Pair> queue = new LinkedList<>();
        queue.add(new Pair(root, 0));
        while (!queue.isEmpty()) {
            Pair currentPair = queue.poll();
            TreeNode currentNode = currentPair.root;
            int currentNodeLevel = currentPair.stateCount;

            lookups.put(currentNodeLevel, currentNode.data);

            if (currentNode.left != null) queue.add(new Pair(currentNode.left, currentNodeLevel + 1));
            if (currentNode.right != null) queue.add(new Pair(currentNode.right, currentNodeLevel + 1));
        }
        for (int vals : lookups.values()) {
            System.out.print(vals + " ");
        }
    }

    private static void findRightViewSpaceOptimized(TreeNode root, int level, ArrayList<Integer> ans) {
        if (root == null) return;
        if (level == ans.size()) {
            ans.add(root.data);
        }
        findRightViewSpaceOptimized(root.right, level + 1, ans);
        findRightViewSpaceOptimized(root.left, level + 1, ans);
    }

    private static void rightViews(TreeNode root) {
//        findRightViewOptimized(root);
        ArrayList<Integer> ans = new ArrayList<>();
        findRightViewSpaceOptimized(root, 0, ans);
        for (int i : ans) System.out.print(i + " ");
    }


    private static void findLeftView(TreeNode root) {
        Queue<Pair> q = new LinkedList<>();
        Map<Integer, Integer> map = new HashMap<>();
        q.add(new Pair(root, 0));

        while (!q.isEmpty()) {
            Pair currentPair = q.poll();
            TreeNode currentNode = currentPair.root;
            int currentLevel = currentPair.stateCount;
            if (!map.containsKey(currentLevel)) {
                map.put(currentLevel, currentNode.data);
            }

            if (currentNode.left != null) q.add(new Pair(currentNode.left, currentLevel + 1));
            if (currentNode.right != null) q.add(new Pair(currentNode.right, currentLevel + 1));
        }
        for (int val : map.values())
            System.out.print(val + " ");
    }

    private static void findLeftViewSpaceOptimized(TreeNode root, int level, ArrayList<Integer> ans) {
        if (root == null) return;
        if (ans.size() == level) ans.add(root.data);
        findLeftViewSpaceOptimized(root.left, level + 1, ans);
        findLeftViewSpaceOptimized(root.right, level + 1, ans);
    }

    private static void leftViews(TreeNode root) {
        findLeftView(root);
        System.out.println();
        ArrayList<Integer> ans = new ArrayList<>();
        findLeftViewSpaceOptimized(root, 0, ans);
        for (int i : ans) System.out.print(i + " ");
    }


    private static void findTopView(TreeNode root) {
        Queue<Pair> q = new LinkedList<>();
        Map<Integer, Integer> map = new TreeMap<>();
        q.add(new Pair(root, 0));
        while (!q.isEmpty()) {
            Pair currentPair = q.poll();
            TreeNode currentNode = currentPair.root;
            int currentBar = currentPair.stateCount;

            if (!map.containsKey(currentBar)) map.put(currentBar, currentNode.data);

            if (currentNode.left != null) q.add(new Pair(currentNode.left, currentBar - 1));
            if (currentNode.right != null) q.add(new Pair(currentNode.right, currentBar + 1));
        }

        for (int i : map.values()) System.out.print(i + " ");
    }

    private static void findBottomView(TreeNode root){
        Queue<Pair> q = new LinkedList<>();
        Map<Integer, Integer> map = new TreeMap<>();
        q.add(new Pair(root, 0));
        while (!q.isEmpty()) {
            Pair currentPair = q.poll();
            TreeNode currentNode = currentPair.root;
            int currentBar = currentPair.stateCount;

            map.put(currentBar, currentNode.data);

            if (currentNode.left != null) q.add(new Pair(currentNode.left, currentBar - 1));
            if (currentNode.right != null) q.add(new Pair(currentNode.right, currentBar + 1));
        }

        for (int i : map.values()) System.out.print(i + " ");
    }

    public static void main(String[] args) {
//        TreeNode root = TreeNode.getInstance2();
//        rightViews(root);
//        leftViews(root);
//        findTopView(root);
//        findBottomView(root);

    }
}

File: Trees/BinaryTrees/ViewsOfTree.cpp
#include <bits/stdc++.h>
using namespace std;

typedef struct TreeNode{
    int val;
    TreeNode* left,*right;

    TreeNode(){
    }

    TreeNode(int val){
        this->val = val;
        left = right = NULL;
    }
}TreeNode;

class Util{
	public:
    static TreeNode* getTree(){
        TreeNode* root = new TreeNode(10);
        root->left = new TreeNode(20);
        root->right = new TreeNode(30);

        root->left->left = new TreeNode(40);
        root->left->right = new TreeNode(50);
        root->left->right->left = new TreeNode(60);
        root->left->right->right = new TreeNode(70);
        return root;
    }

    static void leftViewOfTree(TreeNode* root){
        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()){
            int size = q.size();
            for(int i=0;i<size;i++){
                if(i == 0)  cout<<(q.front())->val<<" ";
                TreeNode* currentNode = q.front();
                q.pop();
                if(currentNode->left != nullptr)    q.push(currentNode->left);
                if(currentNode->right != nullptr)   q.push(currentNode->right);
            }
        }
    }

    static void rightViewOfTree(TreeNode* root){
        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()){
            int size = q.size();
            for(int i=0;i<size;i++){
                if(i == (size - 1))  cout<<(q.front())->val<<" ";
                TreeNode* currentNode = q.front();
                q.pop();
                if(currentNode->left != nullptr)    q.push(currentNode->left);
                if(currentNode->right != nullptr)   q.push(currentNode->right);
            }
        }
    }

    static void topViewOfTree(TreeNode* root){
        map<int,int> lookup;
        queue<pair<TreeNode*,int> > q;
        q.push({root,0});

        while(!q.empty()){
            TreeNode* currentNode = q.front().first;
            int currentIndicator = q.front().second;
            q.pop();
            
            if(lookup.find(currentIndicator) == lookup.end())
                lookup[currentIndicator] = currentNode->val;

            if(currentNode->left != nullptr)    q.push({currentNode->left,currentIndicator-1});
            if(currentNode->right != nullptr)    q.push({currentNode->right,currentIndicator+1}); 
        }

        for(pair<int,int> p : lookup){
            cout<<p.second<<" ";
        }
    }

    static void bottomViewOfTree(TreeNode* root){
        map<int,int> lookup;
        queue<pair<TreeNode*,int> > q;
        q.push({root,0});

        while(!q.empty()){
            TreeNode* currentNode = q.front().first;
            int currentIndicator = q.front().second;
			q.pop();
			
            lookup[currentIndicator] = currentNode->val;

            if(currentNode->left != nullptr)    q.push({currentNode->left,currentIndicator-1});
            if(currentNode->right != nullptr)    q.push({currentNode->right,currentIndicator+1}); 
        }

        for(pair<int,int> p : lookup){
            cout<<p.second<<" ";
        }
    }

};

int main(){
    TreeNode* root = Util::getTree();
    Util::bottomViewOfTree(root);
    
    return 0;
}
